### ISSUE TITLE
task1_ai_ml_model :CBT classification based therapy algorithmic research

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#7]

**Comment by rrupam** [comment]
Deep learning for language understanding of mental health concepts derived from Cognitive Behavioural Therapy


[Here](https://arxiv.org/pdf/1809.00640.pdf#page11) is the relevant paper. 


**Aim of the paper**
To develop a DL algorithm to assess a person based on their post. 

- Establish CBT based ontology  - defines the concepts that the dialogue system can talk about. 
             - Thinking errors - eg. Fortune-telling, emotional reasoning etc.
             - Emotion - eg. Anxiety, anger etc.
             - Situation - eg. Relationship, work etc.
-NLU - identifies ontology items in an input.

**Data**
The corpus consists of 500K written posts that users anonymously posted on the Koko platform. In this set-up, a user anonymously posts something (referred to as the problem) and is prompted to consider their most negative take on the problem (referred to as the negative take). 
A subset of posts was annotated by two psychological therapists using a web annotation tool that we developed. 

**Methodology**
- CNN + GloVe - Each post is passed through the embedding and then CNN, the corresponding -ve take is also passed through a different CNN and then their extracted features are combined in a Fully connected NN to give a binary classification (Why binary?) 
- GRU - Implement skip thought model to generate sentences for similarity calculation. (I’m not entirely sure how this works)

**Conclusion**
We found that classifying thinking errors is a difficult task as suggested by the low inter-annotator agreement. 
 The results suggest that both models significantly outperform a chance classifier for all thinking errors, emotions and situations with CNNGloVe on average achieving better results.

**Concerns**
- Regarding assessement of PTSD condition, does it make sense to have a NLU model that performs okish as compared to a self-assessement that is likely more accurate?
- We will have to look into a possibly different ontology that is specific to PTSD. Does this even exist? Ask Alexis.
- This specific model requires input in the form of POST and NEGATIVE TAKE. Our current dataset does not meet this criterion.

**Technical questions**
- Why is binary classification applied at the end of CNN, when there are more classes?
- What does the GRU model do exactly? It is supposed to generate new thoughts, but how does this help with classification?

**Comment by elateifsara** [comment]
> * It

Binary because they are classifying the feeling either negative or positive.

**Comment by kulsoom-abdullah** [comment]
I read the paper - Deep learning for language understanding of mental health concepts derived from Cognitive Behavioural Therapy

I was first interested in the data they used which was 500K posts in an online platform. These posts were a negative take on a problem, and peers respond with a positive rethink. The responses were not collected for the dataset.

Here is one of the results:
Problem: I really miss my ex! She just stopped talking to me all of the sudden, I don't know what happened
Negative Take: I wont get to talk to her again
Thinking Errors Emotions Situation
True LABELS

Jumping to negative conclusions - Hurt - Relationship
Low Frustration Tolerance - Grief - Relationship

A question I have is this similar to the goal of the project - that is, to take words of someone and detect or classify something related to PTSD and guide them where to get help? I want to make sure I understand. Thanks

**Comment by kulsoom-abdullah** [comment]
How much data do we need to implement deep learning? If we don't have enough, what could we try first?
RELATED WORK: a decision tree used in :
Kathleen Kara Fitzpatrick, Alison Darcy, and Molly Vierhile. 2017. "Delivering cognitive behavior therapy to young adults with symptoms of depression and anxiety using a fully automated conversational agent (woebot): a randomized controlled trial. JMIR mental health, 4(2).
 “inability to converse naturally” 
ok in our case since we don't want it to be like a human

**Comment by elateifsara** [comment]
@kulsoom-abdullah did you read the whole paper? I just found it and was thinking about reading.

**Comment by kulsoom-abdullah** [comment]
> @kulsoom-abdullah did you read the whole paper? I just found it and was thinking about reading it.

I did though I skimmed the methodology part. @rrupam has a good summary posted if you want to focus on certain parts of skip reading the paper.

**Comment by SubmitCode** [comment]
I read the paper as well and I think again as we don't have as much data (allthough we can overcome some obstacles with transfer learning) and because it is quiete a delicate topic I would rather try to build the foundation for a system which allows experts to label and make the system better overtime.

**Comment by nlauchande** [comment]
As suggested by @rrupam  i changed the title .

**Comment by albertlai431** [comment]
Resource on Woebot for CBT!
https://blog.re-work.co/how-are-ai-assistants-transforming-healthcare-woebot-case-study/

**Comment by BottleSpink** [comment]
Topic: 
[[Delivering Cognitive Behavior Therapy to Young Adults with Symptoms of Depression and Anxiety Using a Fully Automated Conversational Agent (Woebot): A Randomized Controlled Trial]](https://mental.jmir.org/2017/2/e19/)
Authors:
             Kathleen Kara Fitzpatrick, Alison Darcy, Molly Vierhile

Objective: 
            To determine the feasibility, acceptability, and preliminary efficacy of a fully automated conversational agent **to deliver a self-help program for college students** who self-identify as having symptoms of anxiety and depression.

Study: 
             The current study compared outcomes from **2 weeks of a CBT-oriented conversational agent (Woebot), or an information control group (National Institute of Mental Health’s [NIMH] ebook)** in a nonclinical college population. 

Hypothesis: 
             Conversation with a therapeutic process-oriented conversational agent would lead to greater improvement in symptoms relative to the information control group.

Conclusion: 
              Conversational agents appear to be a feasible, engaging, and effective way to deliver CBT.

Questionnaire: 
               Web-based versions of the 9-item **Patient Health Questionnaire (PHQ-9), the 7-item Generalized Anxiety Disorder scale (GAD-7), and the Positive and Negative Affect Scale** at baseline and 2-3 weeks later (T2)

CBT framework: 
               Incorporating automated tailoring; **reporting of thoughts, feelings, or behaviors; recommending activities;** provision of mental health information; real-time engagement; **activities explicitly linked to specific reported mood problems**; encouraging non‒technology-based activities; **gamification and intrinsic motivation to engage; reminders to engage;** simple and intuitive interface and interactions; and including links to crisis support services.

Interactions: 
              Each interaction begins with a **general inquiry about the context** (eg, “What’s going on in your world right now?”), and **mood** (eg, “How are you feeling?”) with responses provided as word or emoji images to represent effect at that moment. After gathering mood data, participants are presented with core concepts related to CBT by links to short videos, or by way of short “word games” designed to **facilitate teaching participants about cognitive distortions.**

Disclaimer: 
             The bot also briefly explained CBT and notified the user that while a psychologist was **“keeping an eye on things”** (ie, monitoring), this was not happening in real-time and **thus the service should not be used as a replacement for therapy.** In addition, participants were encouraged to call 911 for emergencies.

Methodology:
1.	A decision tree with suggested responses that also accepted natural language inputs with discrete sections of natural language processing techniques embedded at specific points in the tree to determine to route to subsequent conversational nodes.
2.	Weekly graphs were processed using temporal pattern recognition to provide users with weekly mood description.
3.	For the duration of the study, the **decision tree structure remained the same for each participant** and parameters did not change depending on the participants’ inputs.	
CBT Content: 
Therapeutic process-oriented features: 
Empathic responses, Tailoring, Goal setting, Accountability, Motivation and engagement, Reflection
Assessment and Statistical Technique: 
5-point Likert scale for satisfaction and ANCOVA to compare the treatment groups

Other notes: 
            The number of participants reporting that the bot felt empathic is noteworthy and comments that referred to the bot as “he,” “a friend,” and a “fun little dude” suggest that the perceived source of empathy was Woebot rather than the bot’s developers. This is especially noteworthy since a purposefully robotic name “Woebot” was chosen to emphasize the nonhuman nature of the agent.	

                          Reference paper worthy to look at: https://mental.jmir.org/2016/1/e7/

**Comment by albertlai431** [comment]
@BottleSpink This is great, thanks! 👏 From what I see, this is more of a therapy-based chatbot than a risk assessment?

**Comment by BottleSpink** [comment]
> @BottleSpink This is great, thanks! 👏 From what I see, this is more of a therapy-based chatbot than a risk assessment?

Yes, Albert. CBT based therapy. I went through the paper and gave a brief to task1 team, last week before we decided to go ahead with risk assessment. Just got time to update here in the issue.

**Comment by rrupam** [comment]
@BottleSpink This is great. I will explore this further.

**Comment by elateifsara** [comment]
@BottleSpink I will check it out tomorrow and get back to you. Thank you so much for the review.
